service/chartconfig/v1/resource/chart/update.go

```diff
@@ -2,14 +2,81 @@ package chart
 
 import (
 	"context"
+	"fmt"
 
+	"github.com/giantswarm/chart-operator/service/chartconfig/v1/key"
+	"github.com/giantswarm/microerror"
 	"github.com/giantswarm/operatorkit/framework"
 )
 
 func (r *Resource) ApplyUpdateChange(ctx context.Context, obj, updateChange interface{}) error {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	chartState, err := toChartState(updateChange)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updating chart %s", chartState.ChartName))
+
+	name := key.ChartName(customObject)
+	releaseName := chartState.ReleaseName
+	channel := chartState.ChannelName
+
+	tarballPath, err := r.apprClient.PullChartTarball(name, channel)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.helmClient.UpdateReleaseFromTarball(releaseName, tarballPath)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updated chart %s", chartState.ChartName))
+
 	return nil
 }
 
 func (r *Resource) NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
-	return nil, nil
+	create, err := r.newCreateChange(ctx, obj, currentState, desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	update, err := r.newUpdateChange(ctx, obj, currentState, desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	patch := framework.NewPatch()
+	patch.SetCreateChange(create)
+	patch.SetUpdateChange(update)
+
+	return patch, nil
+}
+
+func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error) {
+	currentChartState, err := toChartState(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredChartState, err := toChartState(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "finding out if the chart has to be updated")
+
+	if currentChartState.ReleaseVersion != "" && desiredChartState.ReleaseVersion != currentChartState.ReleaseVersion {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "the chart has to be updated")
+
+		return &desiredChartState, nil
+	} else {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "the chart does not have to be updated")
+	}
+
+	return &ChartState{}, nil
```

Comment:
Is there a special reason why we want to return an initialized pointer here without any information? I think in such cases we should return nil. The handling of nil is way easier to check against. Getting initialized pointers without any values should be an error. The resource wants to try to do something useful and when it gets something which is actually nothing it cannot do anything. Getting nothing (`nil`) means to do nothing, which is straight forward.

Reply:
+1 I think returning nil is cleaner here.

Reply:
Agree, done PTAL.